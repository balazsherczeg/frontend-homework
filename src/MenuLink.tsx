import DropdownCloser from './DropdownCloser';

type Props = {
  text: string,
  icon: string,
  className?: string,
} & ({ onClick: () => void } | { href: string })

const MenuLink = (props: Props) => {
  const { icon, text, className } = props;

  const handleClick = () => {
    if ("href" in props) {
      window.location.assign(props.href)
    } else {
      props.onClick()
    }
  }

  return (
    <div className={`MenuLink ${className}`}>
      <DropdownCloser>
        <div
          className="MenuLink__inner align-items-vertically-center"
          role="button"
          onClick={handleClick}
        >
          <i className="MenuLink__icon material-icons">
            {icon}
          </i>

          {text}
        </div>
      </DropdownCloser>
    </div>
  )
}

export default MenuLink
