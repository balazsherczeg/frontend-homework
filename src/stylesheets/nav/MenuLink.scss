.MenuLink {
  height: 32px;

  &:not(:last-child) {
    margin-bottom: 4px;
  }
}

.MenuLink__inner {
  @include menu-item;
  @include state-feedback;

  height: 100%;
  width: 100%;
}

.MenuLink__icon {
  padding-right: 8px;
}
