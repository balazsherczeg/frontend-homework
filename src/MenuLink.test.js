import { shallow } from 'enzyme'
import MenuLink from './MenuLink'

const render = (props) => shallow(
  <MenuLink
    icon="phone"
    text="Call me"
    {...props}
  />
)

const handleClickMock = jest.fn()

describe('<MenuLink />', () => {

  it('renders link', () => {
    expect(render({ onClick: handleClickMock })).toMatchSnapshot()
  })
})

describe('<MenuLink /> behaves as a button', () => {
  it('calls handleClickMock on click if onClick prop is provided', () => {
    const component = render({  onClick: handleClickMock })
    component.find('[role="button"]').simulate('click')
    expect(handleClickMock).toHaveBeenCalled()
  })
})

describe('<MenuLink /> behaves as link', () => {
  const { location } = window;

  beforeAll(() => {
    delete window.location;
    window.location = { assign: jest.fn() };
  });

  afterAll(() => {
    window.location = location;
  });

  it('and changes location on click if href prop is provided', () => {
    const component = render({  href: '#aSimpleUrl' })
    component.find('[role="button"]').simulate('click')
    expect(window.location.assign).toHaveBeenCalled()
  })
})
