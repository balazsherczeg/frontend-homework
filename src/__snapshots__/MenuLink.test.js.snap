// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuLink /> renders link 1`] = `
<div
  className="MenuLink undefined"
>
  <Connect(DropdownCloser)>
    <div
      className="MenuLink__inner align-items-vertically-center"
      onClick={[Function]}
      role="button"
    >
      <i
        className="MenuLink__icon material-icons"
      >
        phone
      </i>
      Call me
    </div>
  </Connect(DropdownCloser)>
</div>
`;
